service/storage_service.cc

```diff
@@ -4589,6 +4589,42 @@ future<> storage_service::wait_for_topology_not_busy() {
     }
 }
 
+future<locator::effective_replication_map_ptr> storage_service::get_erm_when_table_has_no_migrations(sstring ks_name, sstring table_name) {
+    lw_shared_ptr<replica::table> t;
+    try {
+        t = _db.local().find_column_family(ks_name, table_name).shared_from_this();
+    } catch (replica::no_such_column_family& e) {
+        throw std::runtime_error(format("Table {}.{} does not exist", ks_name, table_name));
+    }
+    if (!t->uses_tablets()) {
+        throw std::runtime_error(format("Table {}.{} is not a tablet table", ks_name, table_name));
+    }
+    const table_id tid = t->schema()->id();
+    // Invoke group0 read barrier before obtaining erm pointer so that it sees all prior metadata changes
+    auto& local_mm = _migration_manager.local();
+    if (local_mm.use_raft()) {
+        // Trigger read barrier to synchronize schema.
+        co_await local_mm.get_group0_barrier().trigger(local_mm.get_abort_source());
+    }
+
+    locator::effective_replication_map_ptr erm;
+    while (true) {
+        erm = t->get_effective_replication_map();
+        const locator::tablet_map& tmap = erm->get_token_metadata_ptr()->tablets().get_tablet_map(tid);
+        if (!tmap.has_transitions()) {
```

Comment:
Is this guaranteed to ever happen?

Reply:
Also, can't we truncate while migrations are happening? Drop the sstables and memtables on both sides.

Reply:
> Is this guaranteed to ever happen?

I think so, but I am not sure. I will have to look into it. Either way, adding a timeout to `get_effective_replication_map()` would solve that problem. And also the problem of `get_effective_replication_map()` waiting for longer than the truncate timeout allows.

> Also, can't we truncate while migrations are happening? Drop the sstables and memtables on both sides.

I don't understand the question. Isn't the purpose of this change to avoid those situations?

Reply:
> > Is this guaranteed to ever happen?
> 
> I think so, but I am not sure. I will have to look into it. Either way, adding a timeout to `get_effective_replication_map()` would solve that problem. And also the problem of `get_effective_replication_map()` waiting for longer than the truncate timeout allows.
> 

Adding a timeout doesn't solve the problem, because it leaves the user with a failure.

> > Also, can't we truncate while migrations are happening? Drop the sstables and memtables on both sides.
> 
> I don't understand the question. Isn't the purpose of this change to avoid those situations?

The goal is to remove the dependency on having no migrations, because it's not clear to me we can guarantee it happens, and because it adds dependencies between unrelated parts of the system.

get_effective_replication_map_ptr() freezes the topology while it is held. If we hold it, there may be migrations ongoing, but there won't be new migrations. So we can truncate on all known replicas, including migration sources and migration destinations, and this should be sufficient.

Reply:
But without synchronizing with migrations, sstables could be in transit, i.e. being streamed to the destination but not sealed yet.
In this case truncating the destination will not help since we collect only "official" sstables that are already sealed and added to the table.
And deleting those at the source without synchronization would also not help since for file-based migration we take a snapshot of all sstables in the tablet and stream from the snapshot.

Reply:
I guess migration (both types) could synchronize with the truncation.

Easier: synchronize at the end of the tablet migration
Harder: abort the migration and synchronize at the (failed) end

We have exactly the same problem with compaction, and with post-repair work. How is it solved?

Reply:
> I guess migration (both types) could synchronize with the truncation.

That was the idea behind making truncate a tablet transition.

> Easier: synchronize at the end of the tablet migration Harder: abort the migration and synchronize at the (failed) end
> 
> We have exactly the same problem with compaction, and with post-repair work. How is it solved?

For compaction, `database::truncate_table_on_all_shards` stops and disables compaction around `database::truncate`

What do you mean by "post-repair work"?  Off-strategy compaction?  It should be stopped by the above.
View-update generation?  good question.  I won't be surprised if it's serialized well against truncate.

Reply:
> > I guess migration (both types) could synchronize with the truncation.
> 
> That was the idea behind making truncate a tablet transition.
> 
> > Easier: synchronize at the end of the tablet migration Harder: abort the migration and synchronize at the (failed) end
> > We have exactly the same problem with compaction, and with post-repair work. How is it solved?
> 
> For compaction, `database::truncate_table_on_all_shards` stops and disables compaction around `database::truncate`
> 
> What do you mean by "post-repair work"? Off-strategy compaction? It should be stopped by the above. View-update generation? good question. I won't be surprised if it's serialized well against truncate.

Everything that's done after repair completes and makes new sstables available.

Reply:
> Everything that's done after repair completes and makes new sstables available.

That's another reason to make repair a tablet transition. So that we don't have to think about synchronization of crossproduct of N algorithms. 

Currently, and after this PR, active repair won't block truncate, so repair will race with truncation. It's still not really safe.

If truncate was a topology request, we could preempt load balancing. We could also avoid latency by canceling all transitions together with submitting the request. There's no point waiting for them to complete .

Reply:
Closing this conversations, as it relates to the older design.